Nrf radio energy scan for open thread

MarekPorwisz opened the pull request.

Commit:
Bluetooth: controller: split: fix LL/CON/MAS/BV-74-C tx timing

Fix EBQ tests for the Max Tx Time and Max Rx Time parameter.

Signed-off-by: Andries Kruithof <Andries.Kruithof@nordicsemi.no>

Bluetooth: controller: split: fixed for endianness

Added conversion to correct endianness

Signed-off-by: Andries Kruithof <Andries.Kruithof@nordicsemi.no>

avisconti committed:
driver/sensor: lis3mdl: Fix DRDY interrupt

The LIS3MDL sensor provides two different pins for handling
interrupts: the DRDY, that triggers new data sample availability,
and INT, that goes off when data sample exceeds a given threshold.

The driver handled data ready triggers only, which does not
require sensor configuration at all. Moreover a dummy read is
required when the data ready is configured to re-trigger a new
interrupt.

Signed-off-by: Armando Visconti <armando.visconti@st.com>

avisconti committed:
boards/shields: x_nucleo_iks01a1: connect irq to DRDY pin

The LIS3MDL driver handles DRDY interrupt only. So connect
trigger to the DRDY pin instead of the INT pin.

Signed-off-by: Armando Visconti <armando.visconti@st.com>

mtpr-ot committed:
Bluetooth: controller: Prevent LLCP handling of must-expire events

When a must-expire ticker callback is executed, it is important not to
call the LLCP state machine, as the lazy state is unavailable. The code
must rely on the next proper event to call ull_conn_llcp with an updated
non-zero lazy count.

Signed-off-by: Morten Priess <mtpr@oticon.com>

rveerama1 committed:
net: ppp: Fix integer handling coverity issue

Store the time difference value in u32_t variable than
abs() value in signed variable.

Fixes #22912
Coverity CID: 208406

Signed-off-by: Ravi kumar Veeramally <ravikumar.veeramally@linux.intel.com>

ioannisg committed:
doc: release notes: list of ARM SoCs added during v2.2 release cycle

This commit adds the list of ARM SoCs whose support was
added during the Zephyr v2.2 release cycle.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
boards: arm: silabs_exx32: correct Kconfig option prompt

Fix a typo in the option prompt for Kconfig symbol
SOC_SERIES_EFM32GG11B.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

mbolivar-nordic committed:
doc: west: add examples for each topology

Add example west.yml files showing how to build west workspaces of
each type. Split these into separate sections as a result since
they're longer.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
doc: west: update structure page

The main change is that west 0.7 is not a namespace package anymore.
Make some other improvements while here.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
doc: west: 'west installation' is now 'west workspace'

This nomenclature change was done in west 0.7 because it seems to be a
lot easier for people to understand. Propagate it to all the west
documentation.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: split: Fix assert on DLE procedure stall

Fix local initiated Data Length Update procedure from being
stalled when a remote initiates a procedure with instant.

Fixes #23069.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

ioannisg committed:
doc: release notes: add missing ARM Board in v2.2 release notes

Adding Si Labs EFM32 Jake Gecko board in the v2.2 release notes.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

joerchan committed:
Bluetooth: host: Document bt_conn_unref responsibility of caller

Document the need for the caller to call  bt_conn_unref on the
connection object returned from API functions.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Bluetooth: samples: central_hr: Move connected callback

Move the connected callback down together with the disconnected
callback.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Bluetooth: samples: Unref conn object on create connection timeout

Handle connected callback with error status not releasing the default
conn object in central samples. This can happen when the initiator fails
to create the connection within 3 seconds and is canceled by the host.

Also restart the scan role in this case.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

MarekPorwisz committed:
net: ieee802154: Extend radio api to support energy scan feature

Add the Energy Scan capability to the possible capabilities list.
Create new energy scan callback function type to make its usage
more readable.

Signed-off-by: Marek Porwisz <marek.porwisz@nordicsemi.no>

MarekPorwisz committed:
drivers: ieee802154_nrf5: Implement energy scan function of the nrf5 driver

Implement function and necessary callbacks to handle Energy Scan feature
of the nRF radio driver needed by some radio stacks.

Signed-off-by: Marek Porwisz <marek.porwisz@nordicsemi.no>

MarekPorwisz committed:
net: openthread: Implement energy scan in openthread using radio driver api

OpenThread has api to make use of radios energy scan feature.
Implemented this api in zephyr. This allows thread to perform energy
scan when needed.

Signed-off-by: Marek Porwisz <marek.porwisz@nordicsemi.no>

agansari committed:
drivers: eth_mcux phy_setup moved after SMI init

SMI initialization is required to enable PHY communication.
PHY setups needs to run after SMI initialization.

Signed-off-by: Andrei Gansari <andrei.gansari@nxp.com>

Commit:
manifest: EDTT: update manifest to point to new revision in EDTT

For passing the EDTT tests pass/fail criteria needs to be updated,
which is done starting with the revision referenced in the updated
manifest

Signed-off-by: Andries Kruithof <Andries.Kruithof@nordicsemi.no>

rlubos commented on include/net/ieee802154_radio.h:
```diff
@@ -26,14 +26,17 @@ extern "C" {
+	IEEE802154_HW_2_4_GHZ	  = BIT(4), /* 2.4Ghz radio supported */
+	IEEE802154_HW_TX_RX_ACK	  = BIT(5), /* Handles ACK request on TX */
+	IEEE802154_HW_SUB_GHZ	  = BIT(6), /* Sub-GHz radio supported */
+	IEEE802154_HW_ENERGY_SCAN = BIT(7)  /* energy scan supported */
```
s/energy/Energy

rlubos commented on drivers/ieee802154/ieee802154_nrf5.c:
```diff
@@ -145,7 +146,8 @@ static void nrf5_rx_thread(void *arg1, void *arg2, void *arg3)
 {
 	return IEEE802154_HW_FCS | IEEE802154_HW_2_4_GHZ |
-	       IEEE802154_HW_TX_RX_ACK | IEEE802154_HW_FILTER;
+	       IEEE802154_HW_TX_RX_ACK | IEEE802154_HW_FILTER|
```
Missing space before `|`

rlubos commented on drivers/ieee802154/ieee802154_nrf5.c:
```diff
@@ -183,6 +185,23 @@ static int nrf5_set_channel(struct device *dev, u16_t channel)
 }
 
+static int nrf5_energy_scan_start(struct device *dev, u16_t duration,
+				void (*done_cb)(struct device *dev,
```
Alignment. And I guess you could use `energy_scan_done_cb` type here?

rlubos commented on drivers/ieee802154/ieee802154_nrf5.c:
```diff
@@ -183,6 +185,23 @@ static int nrf5_set_channel(struct device *dev, u16_t channel)
+	ARG_UNUSED(dev);
+	if (energy_scan_done == NULL) {
+		energy_scan_done = done_cb;
+		nrf_802154_energy_detection(duration * 1000);
```
We shall check the return value, right?

rlubos commented on drivers/ieee802154/ieee802154_nrf5.c:
```diff
@@ -541,6 +560,29 @@ void nrf_802154_cca_failed(nrf_802154_cca_error_t error)
+{
+	if (energy_scan_done != NULL) {
+		s16_t dbm;
+		energy_scan_done_cb callback = energy_scan_done;
```
You need an empty line after declarations.

MarekPorwisz replied:
energy_scan_done_cb callback = energy_scan_done; is also a declaration, there is a new line after

rlubos replied:
True, lack of highlighting on GH confused me 😕

rlubos commented on drivers/ieee802154/ieee802154_nrf5.c:
```diff
@@ -541,6 +560,29 @@ void nrf_802154_cca_failed(nrf_802154_cca_error_t error)
+		callback(net_if_get_device(nrf5_data.iface), SHRT_MAX);
+	}
+}
+
```
Excess empty line.

MarekPorwisz replied:
removed

rlubos commented on drivers/ieee802154/ieee802154_nrf5.c:
```diff
@@ -556,7 +598,11 @@ static struct ieee802154_radio_api nrf5_radio_api = {
+	.ed_scan = nrf5_energy_scan_start,
+#endif /* CONFIG_NET_L2_OPENTHREAD */
 	.configure = nrf5_configure,
+
```
Excess empty line.

rlubos commented on subsys/net/lib/openthread/platform/radio.c:
```diff
@@ -43,6 +42,15 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
 #define FRAME_TYPE_ACK 0x02
 
+
+enum pending_events_t {
```
No need for `_t` suffix.

rlubos commented on subsys/net/lib/openthread/platform/radio.c:
```diff
@@ -337,11 +430,21 @@ void otPlatRadioSetPromiscuous(otInstance *aInstance, bool aEnable)
+
+	if (!radio_api->ed_scan(radio_dev, energy_detection_time,
+				energy_detected)) {
+		reset_pending_event(pending_events_detect_energy);
```
Pending events were just cleared, is there a need to reset it again? Nonetheless, IMO it would be good to add LOG_ERR here in case we fail to setup the ED procedure.

rlubos replied:
And return an appropriate error code of course.

rlubos replied:
Ok I think I get it now. The problem is with `otPlatRadioEnergyScan` API specification - it simply does not accept the fact that triggering energy detection can fail. The only accepted return values are `OT_ERROR_NONE` and `OT_ERROR_NOT_IMPLEMENTED`. That said, IMO we could leave it as it is, but please add a comment here, why is this done like this.

rlubos commented on subsys/net/lib/openthread/platform/radio.c:
```diff
@@ -309,9 +378,33 @@ int8_t otPlatRadioGetRssi(otInstance *aInstance)
+
+	enum ieee802154_hw_caps radio_caps;
 	ARG_UNUSED(aInstance);
+	__ASSERT(radio_dev,
```
Do we need to check both? Probably just a check for `radio_api` is enough?

rlubos commented on subsys/net/lib/openthread/platform/radio.c:
```diff
@@ -309,9 +378,33 @@ int8_t otPlatRadioGetRssi(otInstance *aInstance)
+	radio_caps = radio_api->get_capabilities(radio_dev);
+	if (radio_caps & IEEE802154_HW_TX_RX_ACK) {
+		/*
+		 * In current state of nrf radio driver ack timeout is not
```
We should not mention nRF radio driver here, it's a generic layer. IMO we should only focus on `OT_RADIO_CAPS_ENERGY_SCAN` in this PR.

rlubos commented on subsys/net/lib/openthread/platform/radio.c:
```diff
@@ -183,7 +228,31 @@ void platformRadioProcess(otInstance *aInstance)
+			radio_api->set_channel(radio_dev,
+					energy_detection_channel);
+
+			if (!radio_api->ed_scan(radio_dev,
```
Something's off here, we just called this function in `otPlatRadioEnergyScan`, why do we do it again?

MarekPorwisz replied:
pending_events_detect_energy event is called when radio_api->ed_scan fails. In such case we schedule the call for later

rlubos commented on subsys/net/lib/openthread/platform/radio.c:
```diff
@@ -58,6 +66,42 @@ static struct ieee802154_radio_api *radio_api;
+{
+	if (dev == radio_dev) {
+		energy_detected_value = max_ed;
+		set_pending_event(pending_events_detect_energy_done);
```
Shouldn't we call `otSysEventSignalPending();` here?

MarekPorwisz replied:
set_pending_event does it

rlubos commented on subsys/net/lib/openthread/platform/radio.c:
```diff
@@ -43,6 +42,15 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
 
+
+enum pending_events_t {
+	pending_events_detect_energy,  /* Requested to start Energy Detection
```
enum values are typically ALL_UPPER_CASE in Zephyr, I have seen no other. Probably enforced by the CI.

rlubos reviewed (commented):
A general comment - your commit messages need to follow specific format, otherwise they'll be rejected by the CI:
https://docs.zephyrproject.org/latest/contribute/index.html#commit-guidelines

tkons commented on drivers/ieee802154/ieee802154_nrf5.c:
```diff
@@ -45,6 +45,7 @@ struct nrf5_802154_config {
 
 static struct nrf5_802154_data nrf5_data;
+static energy_scan_done_cb energy_scan_done;
 
```
= NULL;

MarekPorwisz replied:
not allowed by coding standard

tkons commented on subsys/net/lib/openthread/platform/radio.c:
```diff
@@ -58,6 +66,42 @@ static struct ieee802154_radio_api *radio_api;
+void energy_detected(struct device *dev, s16_t max_ed)
+{
+	if (dev == radio_dev) {
+		energy_detected_value = max_ed;
```
why conversion from s16_t to s8_t

MarekPorwisz replied:
moved conversion to the last possible moment

tkons commented on subsys/net/lib/openthread/platform/radio.c:
```diff
@@ -183,7 +228,31 @@ void platformRadioProcess(otInstance *aInstance)
+	if (event_pending) {
+		otSysEventSignalPending();
 	}
+
```
remove empty line

Commit:
tracing: fixes C++ compilation issue

When tracing_cpu_stats.h is included by C++ file it will not compile
and link correctly due to missing #ifdef __cplusplus directives
in the header file.

Fixes #23072

Signed-off-by: Dariusz Lisik <dariusz.lisik@hidglobal.com>

Commit:
release: Zephyr 2.2.0-rc2

Bumped to v2.2.0-rc2

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

carlescufi committed:
Bluetooth: tester: Add missing log_strdup

Since we enable CONFIG_ASSERT in the tester app, the logging subsystem
will assert if it detects a missing log_strdup. Add the two missing ones
that I could find to properly log transient strings.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

rlubos commented on include/net/ieee802154_radio.h:
```diff
@@ -26,14 +26,17 @@ extern "C" {
  * @{
```
nit: The commit title should begin with `net: ieee802154: `
